### ISSUE TITLE
Conan Support for CSharp DLLs as components

### ISSUE THREAD

**Issue by JoshMusick** [conan-io/conan#15407]
### What is your question?

We have a number of C++ libraries that we need to wrap (using SWIG) which generates a CSharp (or other language interface such as python etc.) interface in the form of a DLL.  I was trying to update a conan recipe to define components of the library to include the csharp dll, however, no matter what I've tried, the test package continues to complain that the package doesn't contain the library:

"-- Looking for package libExample
-- Conan: Component target declared 'libExample'
-- Conan: Component target declared 'libExample_CSharp'
CMake Error at build/msvc-192-x86_64-17-debug/generators/cmakedeps_macros.cmake:67 (message):
  Library 'libExample_CSharp-d' not found in package.  If
  'libExample_CSharp-d' is a system library, declare it with
  'cpp_info.system_libs' property
Call Stack (most recent call first):
  build/msvc-192-x86_64-17-debug/generators/libExample-Target-debug.cmake:61 (conan_package_library_targets)
  build/msvc-192-x86_64-17-debug/generators/libExampleTargets.cmake:26 (include)
  build/msvc-192-x86_64-17-debug/generators/libExampleConfig.cmake:16 (include)
  CMakeLists.txt:11 (find_package)

-- Configuring incomplete, errors occurred!"

The CMakeList.txt line 11 is:
 find_package(libExample REQUIRED)

I've looked in the package directory and the libExample_CSharp-d.dll is present in the bin directory.  Additionally, when we tell Conan to use the library generated cmake config files, it works, but I was hoping to let Conan define/provide the cmake configuration files as we are doing for our regular C++ packages.  

My question is mostly regarding, is there support for CSharp DLLs as libraries from Conan? And if so, am I simply not defining the component properties correctly?  Below is a pruned down version of the conanfile.py for our example library.

```
class libexampleConan(ConanFile):
    name = "libexample"
    target_name = "libExample"
    
    # Binary configuration
    settings = "os", "compiler", "build_type", "arch"
    options = {
        "shared": [True, False], 
        "fPIC": [True, False],
        "swig_csharp": [True,False],
        "swig_python": [True,False]
    }
    default_options = {
        "shared": True, 
        "fPIC": True,
        "swig_csharp": False,
        "swig_python": False
    }
        
    package_type = "library"

    def layout(self):
        cmake_layout(self)
        
	    # ... 
        
    def build(self):                
        cmake = CMake(self)
        cmake.configure()
        if self.options.swig_csharp: #or self.options.swig_python:
            if self.options.swig_csharp:
                cmake.build( target=f"{self.target_name}_SWIG_CSharp") # generate the .cs files
            
            # Now, enable all final targets to be built
            cmakeVars = { "BUILD_FINAL_SWIG_TARGETS": "True" }
            cmake.configure(variables=cmakeVars) # configuring the CSharp cmakelists (to use the generated .cs files)            
        cmake.build() # final build of everything
        
    def package(self):        
        cmake = CMake(self)
        cmake.install()
        # this will remove the installed configuration files, and use Conan files instead
        #rmdir(self, os.path.join(self.package_folder, "lib", "cmake"))        

    def package_info(self):
        use_cmakeconfigs = True
        
        ext = ""
        if self.settings.os == "Linux":
            ext = ".so" if self.options.shared else ".a"
        suffix = "-d" if self.settings.build_type == "Debug" else ""
        # this assumes a "-d" postfix for your debug build i.e. 
        # libExample-d.dll vs libExample.dll for debug vs release
		
        if use_cmakeconfigs:        
			# This is currently the only way the test_package works
            self.cpp_info.set_property("cmake_find_mode", "none")
            self.cpp_info.builddirs.append(os.path.join("lib","cmake"))
            return                     
        
        # Define the properties for the base component (the C++ library)
        tarName = self.target_name
        self.cpp_info.set_property("cmake_file_name", f"{tarName}")        
        self.cpp_info.set_property("cmake_target_name", f"{tarName}")
                       
        self.cpp_info.components[f"{tarName}"].libs = [f"{tarName}{suffix}{ext}"]
        self.cpp_info.components[f"{tarName}"].set_property("cmake_file_name", f"{tarName}")
        self.cpp_info.components[f"{tarName}"].set_property("cmake_target_name", f"{tarName}")
            
        if not self.options.shared or self.settings.os == "Linux":
            self.cpp_info.components[f"{tarName}"].defines.append(f"LIBEXAMPLE_STATIC_DEFINE")
                        
        if self.options.swig_csharp:
            # Define the properties for the SWIG_CSharp component 
            tarName = f"{self.target_name}_SWIG_CSharp"
            self.cpp_info.components[f"{tarName}"].libs = [f"{tarName}"]            
            self.cpp_info.components[f"{tarName}"].set_property("cmake_file_name", f"{tarName}")
            self.cpp_info.components[f"{tarName}"].set_property("cmake_target_name", f"{tarName}")
                    
			# Define the properties for the CSharp Component
            tarName = f"{self.target_name}_CSharp"
            self.cpp_info.components[f"{tarName}"].libs = [f"{tarName}{suffix}"]
            self.cpp_info.components[f"{tarName}"].set_property("cmake_file_name", f"{tarName}")
            self.cpp_info.components[f"{tarName}"].set_property("cmake_target_name", f"{tarName}")
```

Thanks in advance!
Josh

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by SpaceIm** [comment]
> ```
> "-- Looking for package libExample
> -- Conan: Component target declared 'libExample'
> -- Conan: Component target declared 'libExample_CSharp'
> CMake Error at build/msvc-192-x86_64-17-debug/generators/cmakedeps_macros.cmake:67 (message):
> Library 'libExample_CSharp-d' not found in package. If
> 'libExample_CSharp-d' is a system library, declare it with
> 'cpp_info.system_libs' property
> Call Stack (most recent call first):
> build/msvc-192-x86_64-17-debug/generators/libExample-Target-debug.cmake:61 (conan_package_library_targets)
> build/msvc-192-x86_64-17-debug/generators/libExampleTargets.cmake:26 (include)
> build/msvc-192-x86_64-17-debug/generators/libExampleConfig.cmake:16 (include)
> CMakeLists.txt:11 (find_package)
> 
> -- Configuring incomplete, errors occurred!"
> 
> The CMakeList.txt line 11 is:
> find_package(libExample REQUIRED)
> ```
> 
> I've looked in the package directory and the libExample_CSharp-d.dll is present in the bin directory.

`CMakeDeps` doesn't care about DLL files (supposed to be in bin folder), it only cares about static lib or import lib of DLL (supposed to be in lib folder). bin folders of requirements are propagated in run environment (`VirtualRunEnv`) so that DLL or executables in these folders can be found at runtime, but DLL are not used by linkers (only import lib of DLL).
Correct me if I'm wrong, but SWIG just generates a DLL without an import lib, and a bunch of cs files exposing C# wrapper code and allowing to dynamically load this DLL at runtime in a C# application, so it doesn't make sense to set `cpp_info.libs` because you don't have an import lib in lib folder.

> Additionally, when we tell Conan to use the library generated cmake config files, it works, but I was hoping to let Conan define/provide the cmake configuration files as we are doing for our regular C++ packages.

You can achieve this in `package_info()` of recipes for which you don't want CMakeDeps to generate anything. Assuming your CMake config file has been installed in `lib/cmake/libname` folder:

```python
def package_info():
    self.cpp_info.set_property("cmake_find_mode", "none")
    self.cpp_info.builddirs.append(os.path.join("lib", "cmake", "libname"))
```

With this updated `package_info()`, when this recipe is a requirement, `CMakeDeps` doesn't generate a CMake config file for this library anymore, and `CMakeToolchain` appends this `lib/cmake/libname` folder to `CMAKE_PREFIX_PATH`, so that CMake config file in package folder can be found by `find_package()`.

**Comment by memsharded** [comment]
Hi @JoshMusick 

Thanks for your questions.

The ``cpp_info`` is defined mostly for build-time information, directories with includes and directories with libraries that must be used at link-time.

Dynamic .dll libraries in Windows, without an import library are not used at link time. If that is the case, then defining ``self.cpp_info.libdirs = []``, ``self.cpp_info.libs = []`` and ``self.cpp_info.bindirs = ["bin"]`` (if dlls are in the bin folder), should be the way to go, and then the ``VirtualRunEnv`` will generate the .bat files that define the PATH env-var so those dlls are found at runtime. Does this make sense?

**Comment by JoshMusick** [comment]
Hi @memsharded 

Yes, your response does make sense.  When I switched to the "self.cpp_info.libdirs = []" it stopped complaining about the csharp lib not being found.  In most of our consumer recipes, we are then copying the .dlls into the application runtime directories (since we can't always guarantee the conanrunenv.bat will be ran.  But this isn't an issue for us.  

My ultimate solution ended up being to define the C++ package properties with conan, and use the CMake generated configs for the CSharp DLLs (so not define the CSharp properties with Conan) rather than trying to define them all as components (which was my original goal).  This required the removal of only the C++ library generated files during the package step.

Otherwise, on the consumer side we ended up needing to "find_package(...)" and then set the VS_DOTNET_REFERENCE ... for the CSharp library.  When using the cmake generated configs, we only need to do a target_link_libraries for the CSharp library dependency.  

Thanks again for the help!  This ticket can be closed.
v/r
Josh
